### ISSUE TITLE
TMDB requests time out during bulk add

### ISSUE THREAD

**Issue by shushry** [Radarr/Radarr#743]
In my specific case, I'm attempting to use the bulk-import feature on nearly 7k titles.     Setting pagination for 1000, the page loads in about 3 minutes in the browser.   I select "all" and then "import selected".
The browser returns, but I'm unable to browse to any other page in Radarr (presumably, it's "doing things")
Tailing the log, indicates that nearly 1,000 titles have been added to the Movies table (confirmed in the sqlite db) - and now it's trying to pull metadata from tmdb.

Seeing continuous "timeout" errors as Radarr tries to pull data from tmdb for images.    Example snippet:

(repeats like this for all the other titles in the first page of import)

Looks like maybe too many requests at once?    This first "page" of the bulk import has been chewing on this for about 3 hours now.

Manually doing a "curl" or "wget" on the requested tmdb jpg URL succeeds from the system in question - so doesn't appear to be a network plumbing issue.

------
17-2-14 02:09:34.4|Warn|HttpClient|Failed to get response from: http://image.tmdb.org/t/p/original/6oSKj6diSeYRykWD9zWFXMDxCWt.jpg The request timed out
17-2-14 02:09:34.4|Warn|MediaCoverService|Couldn't download media cover for [tt2304459][23 Blast]. The request timed out

[v0.2.0.344] System.Net.WebException: The request timed out
  at System.Net.WebClient.DownloadFile (System.Uri address, System.String fileName) [0x000c9] in <5071a6e4a4564e19a2eda0f53e42f9bd>:0 
  at System.Net.WebClient.DownloadFile (System.String address, System.String fileName) [0x00019] in <5071a6e4a4564e19a2eda0f53e42f9bd>:0 
  at (wrapper remoting-invoke-with-check) System.Net.WebClient:DownloadFile (string,string)
  at NzbDrone.Common.Http.HttpClient.DownloadFile (System.String url, System.String fileName) [0x00056] in <fc0cc8da29884ba28b47c33435528961>:0 

17-2-14 02:09:34.4|Warn|HttpClient|Failed to get response from: http://image.tmdb.org/t/p/original/yZSdJNUfbK5y7mT80PwWGgvXkXn.jpg The request timed out
17-2-14 02:09:34.4|Warn|MediaCoverService|Retrying for the 1. time in ten seconds.
17-2-14 02:09:34.4|Warn|MediaCoverService|Couldn't download media cover for [tt1727770][Absolutely Anything]. The request timed out

[v0.2.0.344] System.Net.WebException: The request timed out
  at System.Net.WebClient.DownloadFile (System.Uri address, System.String fileName) [0x000c9] in <5071a6e4a4564e19a2eda0f53e42f9bd>:0 
  at System.Net.WebClient.DownloadFile (System.String address, System.String fileName) [0x00019] in <5071a6e4a4564e19a2eda0f53e42f9bd>:0 
  at (wrapper remoting-invoke-with-check) System.Net.WebClient:DownloadFile (string,string)
  at NzbDrone.Common.Http.HttpClient.DownloadFile (System.String url, System.String fileName) [0x00056] in <fc0cc8da29884ba28b47c33435528961>:0 

------

**Comment by onedr0p** [comment]
7k movies eek! Radarr I don't think is designed to handle that much. Load times would be so slow. I think tmdb may be timing out your requests.

**Comment by shushry** [comment]
Why would 7k, or even 70k be a performance bottleneck?   Radarr should only be handling a few items at at time in it's processes anyway.   In fact, in this case, we're only talking about the first 1000 items (I'd venture to say 1000 items is pretty fair game for anyone's collection) and this issue above is occurring immediately.

By contrast, my Sonarr installation manages a collection of 46k+ episodes across 445 series.   No performance issues there....and since Radarr is a fork of that......

**Comment by Snorflake** [comment]
I have ~150 movies and mine is also just hanging on "Loading search results from TheTVDB for your movies, this may take a few minutes. "

**Comment by galli-leo** [comment]
And yes it takes a long time to fetch all metadata for 1000 movies. Not much we can do about that, you just have to be patient.

**Comment by Snorflake** [comment]
There are no logs for me, and I was waiting on that screen for about 20 minutes (for 156 movies). Does the size matter? All up it's about 1.1TB.

**Comment by galli-leo** [comment]
@Snorflake Size should not matter. Was this when adding them or when just opening the folder?

**Comment by Snorflake** [comment]
@galli-leo Fresh install, went to:

- Add Movies
- Import Existing
- Select folder

And then I'm just waiting on a it to fetch from TVDB (I suppose, there aren't any logs.)

**Comment by galli-leo** [comment]
@Snorflake Yeah, it can take some time depending on your connection. If you are unsure, you can enable trace logging and take a look at the log file. It should show activity there.

**Comment by Snorflake** [comment]
How do I go about doing that? I have a 50/20 conneciton.

**Comment by galli-leo** [comment]
@Snorflake Settings ->General -> Log Level. Then restart Radarr. Then retry bulk import. Open a new tab with Radarr. Go to System -> Logs -> Files -> newest file.

**Comment by Snorflake** [comment]
@galli-leo It's telling me the database is locked (Error 5).

**Comment by shushry** [comment]
Ok, A bit more info here:
 I can successfully add a page of 500  at a time into the bulk importer without TMDB timeouts.  If I select a page of 1000, I'll start to see timeouts listed in the debug log to tmdb.    I'm not sure where the exact number of items is that causes it though.

In my testing, Radarr will pull in titles from bulk import at a rate of about 4-5 per second., regardless of the pagination size.   Sometimes faster.  Perfectly tolerable with even a list of 500.

In this testing I've discovered another issue though that I'll post separately that causes the bulk importer to stop cold and report nothing back - making one believe it's 'frozen' when it simply gave up.

**Comment by shushry** [comment]
I'm able to successfully import blocks of titles in pages (up to 500 even) if I have identified the "problem" titles and removed them - that generate the sql insert error.   Can't yet identify exactly _what_ that error is as  the trace log gets truncated.

**Comment by 2wenty2wo** [comment]
I'm also having this exact same issue. It just hangs on loading results.

**Comment by shushry** [comment]
A bit more here.    The MediaCover service, which retrieves the title poster jpg's etc seems to be what's timing out.  I previously reported that I could do 500 at a time page size - but this is no longer the case.   I can do a block (page) of 100 successfully - watching the trace log it spins through these 100 in about 5 minutes.   Setting the batch/page size to 500, gets the process running, but after what I deem to be about 200 or so I start to see timeouts on requests with the mediacover service.    I can't tell if requests are just continuing to be retried, and eventually will work, or if i'm "blocked" at this point, or what's going on.  Seems to be too many requests in bulk to tmdb.   Anyhoo, the bulk import process won't complete (and requests to other pages in radarr will spin...) until this whole thing finishes, so it gets stuck here.

**Comment by galli-leo** [comment]
@shushry Well it does retry. Have you waited until it finished? Since the media cover service is only activated after all movies were added to the db, you can also try and navigate to the movies section. There you should find all your imported movies.

**Comment by shushry** [comment]
I haven't waited until it finished because it never does - not within a reasonable amount of time anyway.  Once the condition occurs, Radarr seems to indefinitely loop through retries every few minutes and never succeeding.  Almost as if something is "locked up".   I'm only able to recover from the condition by stopping and starting Radarr again.

As an example, in my log snip above, you can see that it times out on the title "23 Blast".  If I tail the trace log under this condition, I'll see that same entry fly by every 2 or 3 minutes.  I've let this run as long as an hour.   Surely it could get a successful retry in by then if it was possible.

Here's a way I'm able to duplicate this issue if you want to try it:

Create a directory tree of many titles - without media.   I'm able to (and it makes sense that you can) bulk add titles with properly named folders, without actually any media.   Example:

testfolder\
  Movie 1 (2001)\
  Movie 2 (1999)\
  Movie 3 (2010)\
  Movie n (xxxx)\

..and so on.

You'll want at least 500 for the test, but several thousand makes sense  and in my case I have over 7k.   I can provide a directory tree for testing if you wish.

1)  Bulk add the directory, and choose a page size of 100.   Process that.   All seems well, and it runs for me in about 5 minutes to completion.
2)  Now bulk add a page size of 500.    Same rate of processing seems to occur, until about 200 or 250 titles or so....then we start to see the "request timed out" in the trace log.   

Again, this never recovers - at least not within an hour (as long as I've tested).   Also - once the condition occurs, I'm not able to navigate to ANY other part of the Radarr application.  Requests will just hang and spin in the browser. The timeout requests will cycle through the trace log however, every few minutes.

Could this be some kind of resource issue, where Radarr is just out of...whatever...and hangs until you restart?

Also - I can confirm, that ALL the titles within the scope of the page (500, say) DO get added to the database (sqlite command line confirmed).  The task just never finishes / recovers / fails.

Happy to provide any other data or scenarios here that are useful at request.    

Also as  a side note, what I'd "really" like to see here, is just the ability to point radarr at a folder and walk away allowing it to chew in the background, the entire directory.  Having to paginate takes away from the "bulk/batch" idea in my opinon and belongs more in the feature next to it (i.e, "Import Existing Movies On Disk").    Of course such a bulk/batch operation assumes a lot on the part of the end user - that they have a folder being added that contains something properly formatted.

And of course - thanks to all contributing to this project - anxious to see this really get off the ground!

**Comment by galli-leo** [comment]
I still think you should be able to check if all movies matched correctly. If you want to have all your movies on one page, you can just edit the html of the pagesize selector and add a 10000 entry. I will try to reproduce this once I have time though.

**Comment by galli-leo** [comment]
@shushry Do you still have issues with the bulk importer? Also for your suggestions I would open a new feature request (since it's not really a bug).

**Comment by shushry** [comment]
Yep thanks for the followup.  It's been working great, and I'm doing them in batches of 100 at a time.  I haven't tested a list longer than about 400 at a time though.  

On April 15, 2017 9:10:21 AM CDT, Leonardo Galli <notifications@github.com> wrote:
>@shushry Do you still have issues with the bulk importer? Also for your
>suggestions I would open a new feature request (since it's not really a
>bug).
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/Radarr/Radarr/issues/743#issuecomment-294295691

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by xxtensazenxx** [comment]
sorry to re-open. I have 640 movies and trying to bulk import, and this issue is still persisting for me. It only happens it seems when you have a above average library. When setting up radarr with my small library i never experienced this. It keeps going in loops. 

Version- 0.2.0.1293
Mono Version - 4.6.2
debug logs- https://drive.google.com/file/d/1rC7fdVsYfbRMbWpVAW440W53MocX5Qso/view?usp=sharing

**Comment by xxtensazenxx** [comment]
Fixed the issue, Updated the mono repo then updating mono thereafter.
